File: LeetCodeProblemsGo/LeetCode2044CountNumberOfMaximumBitwiseORSubsets/LeetCode2044CountNumberOfMaximumBitwiseORSubsets.go
package LeetCode2044CountNumberOfMaximumBitwiseORSubsets

func Backtrack(nums []int, length, targetOrProduct, currentOr, startIndex int, result *int) {
	if currentOr == targetOrProduct {
		*result++
	}
	for i := startIndex + 1; i < length; i++ {
		Backtrack(nums, length, targetOrProduct, currentOr|nums[i], i, result)
	}
}
func countMaxOrSubsets(nums []int) int {
	orProduct := 0
	for _, num := range nums {
		orProduct |= num
	}
	result := 0
	Backtrack(nums, len(nums), orProduct, 0, -1, &result)
	return result
}

File: LeetCodeProblemsCpp/LeetCodeProblemsCpp/Problems/LeetCode2044CountNumberOfMaximumBitwiseORSubsets.cpp
#include "LeetCode2044CountNumberOfMaximumBitwiseORSubsets.h"
using namespace std;

void Backtrack(vector<int>& nums,int targetOrProduct,int length,int startIndex,int orAggregate,int & result)
{
    if (orAggregate==targetOrProduct)
    {
        result++;
    }
    for (int i = startIndex+1; i < length; i++)
    {
        Backtrack(nums,targetOrProduct,length,i,orAggregate|nums[i],result);
    }
        
}
int LeetCode2044CountNumberOfMaximumBitwiseORSubsets::countMaxOrSubsets(vector<int>& nums)
{
    int orAggregate{};
    for (const auto& num : nums)
    {
        orAggregate |= num;
    }
    int result{};
    Backtrack(nums,orAggregate,nums.size(),-1,0,result);
    return  result;
}
